cli/cmd/test_connection.go

```diff
@@ -171,34 +185,146 @@ func testConnection(cmd *cobra.Command, args []string) error {
-		} else {
-			log.Info().Str("source", sources[i].VersionString()).Msg("Initialized source")
+			allErrors = errors.Join(allErrors, fmt.Errorf("failed to test source %v: %w", sources[i].VersionString(), err))
+			continue
 		}
+		testResult.pluginRef = sources[i].VersionString()
+		testResult.pluginKind = "source"
+		testConnectionResults = append(testConnectionResults, *testResult)
 	}
 	for i, client := range destinationClients {
 		pluginClient := plugin.NewPluginClient(client.Conn)
-		log.Info().Str("destination", destinations[i].VersionString()).Msg("Initializing destination")
-		err := initPlugin(ctx, pluginClient, destinations[i].Spec, false, invocationUUID.String())
+		log.Info().Str("destination", destinations[i].VersionString()).Msg("Testing destination")
+		testResult, err := testPluginConnection(ctx, pluginClient, destinations[i].Spec)
 		if err != nil {
-			initErrors = append(initErrors, fmt.Errorf("failed to init destination %v: %w", destinations[i].VersionString(), err))
-		} else {
-			log.Info().Str("destination", destinations[i].VersionString()).Msg("Initialized destination")
+			allErrors = errors.Join(allErrors, fmt.Errorf("failed to test destination %v: %w", destinations[i].VersionString(), err))
+			continue
 		}
+		testResult.pluginRef = destinations[i].VersionString()
+		testResult.pluginKind = "destination"
+		testConnectionResults = append(testConnectionResults, *testResult)
 	}
 
-	allErrors := errors.Join(initErrors...)
 	status := cloudquery_api.SyncTestConnectionStatusCompleted
 	if allErrors != nil {
 		status = cloudquery_api.SyncTestConnectionStatusFailed
 	}
-	updateSyncTestConnectionStatus(ctx, log.Logger, status)
+	updateSyncTestConnectionStatus(context.Background(), log.Logger, status, testConnectionResults...)
+
+	log.Info().Any("testresults", testConnectionResults).Msg("Test connection completed")
+
+	maxLength := 0
+	for _, testResult := range testConnectionResults {
+		if len(testResult.pluginRef) > maxLength {
+			maxLength = len(testResult.pluginRef)
+		}
+	}
+
+	connFailures := make([]testConnectionResult, 0, len(testConnectionResults))
+	for i, testResult := range testConnectionResults {
+		if i == 0 {
+			cmd.Println("Test Connection Results")
+		}
+		cmd.Printf("%-12s %s %s", testResult.pluginKind, testResult.pluginRef, strings.Repeat(" ", maxLength-len(testResult.pluginRef)+1))
+		if testResult.Success {
+			cmd.Println("Success")
+			continue
+		}
+		cmd.Println("Failure:", testResult.FailureCode, "\t", testResult.FailureDescription)
+		connFailures = append(connFailures, testConnectionResults[i])
+	}
+
+	if len(connFailures) > 0 {
+		allErrors = errors.Join(allErrors, &testConnectionFailures{failed: connFailures})
+	}
 
 	return allErrors
 }
+
+type testConnectionFailures struct {
+	failed []testConnectionResult
+}
+
+func (*testConnectionFailures) Error() string {
+	// The errs are already shown in the console, so we don't need to show them again here
```

Comment:
Are the errors shown when you do `--log-console` too?

Reply:
Yes, now better, with proper `plugin_ref` etc.
```
2024-05-17T14:33:56Z INF Loading spec(s) args=["./pgrpc.yml"] invocation-id=dafde1e3-5e32-4fad-8fb5-c7ad7a885284 module=cli
2024-05-17T14:33:56Z INF Testing destination destination="pg (grpc@localhost:7778)" invocation-id=dafde1e3-5e32-4fad-8fb5-c7ad7a885284 module=cli
2024-05-17T14:33:56Z INF Test connection completed invocation-id=dafde1e3-5e32-4fad-8fb5-c7ad7a885284 module=cli testresults=[{"failure_code":"OTHER","failure_description":"rpc error: code = Internal desc = failed to init plugin: failed to initialize client: failed to get current database: failed to connect to `host=localhost user=postgres database=cq`: dial error (dial tcp 127.0.0.1:15432: connect: connection refused)","plugin_kind":"destination","plugin_ref":"pg (grpc@localhost:7778)","success":false}]
Error: at least one test connection failed
2024-05-17T14:33:56Z ERR exiting with error error="at least one test connection failed" invocation-id=dafde1e3-5e32-4fad-8fb5-c7ad7a885284 module=cli
```